Mathlib/MeasureTheory/Function/UnifTight.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2023 Igor Khavkine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Igor Khavkine
+-/
+import Mathlib.MeasureTheory.Function.LpSpace
+
+/-!
+# Uniform tightness
+
+This file contains the definitions for uniform tightness for a family of Lp functions.
+It is used as a hypothesis to the version of Vitali's convergence theorem for Lp spaces
+that works also for spaces of infinite measure.
+
+## Main definitions
+
+* `MeasureTheory.UnifTight`:
+  A sequence of functions `f` is uniformly tight in `L^p` if for all `ε > 0`, there
+  exists some measurable set `s` with finite measure such that the Lp-norm of
+  `f i` restricted to `sᶜ` is smaller than `ε` for all `i`.
+
+# Main results
+
+* `MeasureTheory.unifTight_finite`: a finite sequence of Lp functions is uniformly
+  tight.
+
+## Tags
+
+uniform integrable, uniformly tight, Vitali convergence theorem
+-/
+
+
+namespace MeasureTheory
+
+open Classical Set Filter Topology MeasureTheory NNReal ENNReal
+
+variable {α β ι : Type*} {m : MeasurableSpace α} {μ : Measure α} [NormedAddCommGroup β]
+
+
+
+section UnifTight
+
+/- This follows closely the `UnifIntegrable` section
+from `MeasureTheory.Functions.UniformIntegrable`.-/
+
+variable {f g : ι → α → β} {p : ℝ≥0∞}
+
+
+/-- A sequence of functions `f` is uniformly tight in `L^p` if for all `ε > 0`, there
+exists some measurable set `s` with finite measure such that the Lp-norm of
+`f i` restricted to `sᶜ` is smaller than `ε` for all `i`. -/
+def UnifTight {_ : MeasurableSpace α} (f : ι → α → β) (p : ℝ≥0∞) (μ : Measure α) : Prop :=
+  ∀ ⦃ε : ℝ≥0∞⦄, 0 < ε → ∃ s : Set α, μ s ≠ ∞ ∧ ∀ i, snorm (sᶜ.indicator (f i)) p μ ≤ ε
```

Comment:
This was Yury's design decision (cf [this commit](https://github.com/leanprover-community/mathlib4/pull/9163/commits/d4b82684179c3fd3eeba5883fb1dfea98c240c19) and [comment](https://github.com/leanprover-community/mathlib4/pull/9163#issuecomment-1940535054)). I was originally following the analogy with `UnifIntegrable` that had `⦃ε : ℝ⦄`. Changing this will create some propagation.

Reply:
@urkud, can we have some background explanation?

Reply:
I think that we should have both versions. I don't care which one is the definition and which one is a lemma. Probably, I was wrong and the `NNReal` version should be the definition.

Reply:
I think that's a good idea

Reply:
> I think that's a good idea

I don't know. This sounds like a doubling of the API in size, for both versions of the uniform tightness condition. Is there any example in Mathlib where something like that is actually done (in an economical way, if possible)?

Reply:
Why would it double the size of the API? All it's going to add is a lemma saying that the definitions in terms of `NNReal` and `ENNReal` agree.

Reply:
Sorry, I'm probably missing the point again. If there are multiple versions of `UnifTight` (`\eps` being `Real`, `NNReal` or `ENNReal`), then what is the version to be used in any lemma/theorem that relies on `UnifTight`? Naively, one would add all the corresponding versions, increasing the API by a factor of 2 or 3.

If no that, then one version of `UnifTight` should be primary and only that one used in the API. So which version should that be?

Reply:
When several definitions A and B are equivalent, you don't write `def A`, `def B`, `lemma _ : A \iff B` but `def A`, `lemma _ : A \iff B`.

Here the definition should be in terms of `NNReal` and you can have a lemma stating the equivalence with the `ENNReal` version.

Reply:
OK, one last comment before the decision is made. The type of `ε` already came up already early on (cf [this comment](https://github.com/leanprover-community/mathlib4/pull/9163/files#r1493852124)). The one argument I can see in favor of `ε : ENNReal` is that in all the proofs `ε` is always compared with `ENNReal` because that is the target type of [docs#MeasureTheory.snorm](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/LpSeminorm/Basic.html#MeasureTheory.snorm), which makes some internal manipulations easier.

The tradeoffs are:
* `ε : Real`: mathematically natural, annoying casts in proofs
* `ε : NNReal`: still mathematically natural, still some (but less annoying) casts in proofs
* `ε : ENNReal`: always check `∞` as a special case (unnatural?), no casts in proofs

So, final decision?

Reply:
I already knew this tradeoff, so I'm happy with `NNReal` 😁